### ISSUE TITLE
Question: Why is `delta_score` not flipped for the `low` lambda in the lambdarank objective?

### ISSUE THREAD

**Issue by gabrielwomark** [microsoft/LightGBM#4357]
Hello! I have a question about the implementation of the lambdarank objective function. 
[here](https://github.com/microsoft/LightGBM/blob/bab58d0e90b2ee93d7e0656957e73e92d46554bc/src/objective/rank_objective.hpp#L193) We see that 
`const double delta_score = high_score - low_score;`
And then [p_lambda gets set to the sigmoid of this value](https://github.com/microsoft/LightGBM/blob/bab58d0e90b2ee93d7e0656957e73e92d46554bc/src/objective/rank_objective.hpp#L206)
But then when the lambdas are updated for both the high and low indices, p_lambda is the same. Wouldn't delta_score have to be multiplied by a negative before being passed to the sigmoid function and used to update the low lambda on [line 211] (https://github.com/microsoft/LightGBM/blob/bab58d0e90b2ee93d7e0656957e73e92d46554bc/src/objective/rank_objective.hpp#L211)?

**Comment by shiyu1994** [comment]
@gabrielwomark Thanks for using LightGBM! Though `p_lambda` is the same when updating the gradients (which is `lambdas` in the function `GetGradientsForOneQuery` for `lambdamart`) of high and low indices, but the sign is different. In
https://github.com/microsoft/LightGBM/blob/bab58d0e90b2ee93d7e0656957e73e92d46554bc/src/objective/rank_objective.hpp#L211
it is minus, and in
https://github.com/microsoft/LightGBM/blob/bab58d0e90b2ee93d7e0656957e73e92d46554bc/src/objective/rank_objective.hpp#L213
it is plus.
Suppose the score for high relevance item is `s1` and the score for low relevance item is `s2`, then in the loss function for `lambdamart`, we have the term `ln (sigmoid(s1 - s2))`.

The delta score should not be multiplied by a negative when updating low lambda. Because in the loss (actually, a pseudo loss for lambdamart), we consider only the pairs `(i, j)` such that ground truth relevance of `i` is higher than `j`. In that case, the loss contains only terms like `sigmoid(s_i - s_j)`. Take the derivative, and we will found that the derivatives of `s_i` and `s_j` only differs by a negative sign. The term `s_i - s_j` in the sigmoid function of the derivatives is the same.

For more detailed information about how the gradients and hessians are calculated for lambdamart, you may refer to
https://www.microsoft.com/en-us/research/uploads/prod/2016/02/MSR-TR-2010-82.pdf

If you have any further problem, please feel free to post here.

**Comment by gabrielwomark** [comment]
> The delta score should not be multiplied by a negative when updating low lambda. Because in the loss (actually, a pseudo loss for lambdamart), we consider only the pairs (i, j) such that ground truth relevance of i is higher than j. In that case, the loss contains only terms like sigmoid(s_i - s_j). Take the derivative, and we will found that the derivatives of s_i and s_j only differs by a negative sign. The term s_i - s_j in the sigmoid function of the derivatives is the same.

I realized this point after staring at the original LambdaRank paper for a while haha. But thank you for the response! It's a lot clearer than the explanation I gave to my colleagues!
